gptfixer/gpt.c

```diff
@@ -0,0 +1,786 @@
+    if (last_end_byte_offset > start_of_backup_gpt_table) {
+        errx(1,
+             "Cannot construct new GPT: Partition %" PRIu32 " ends at byte offset 0x%" PRIX64
+             ", but partition table uses space after 0x%" PRIX64,
+             partition_that_ends_last + 1, last_end_byte_offset, start_of_backup_gpt_table);
+    }
+    uint64_t first_usable_byte;
+    CHECK(!__builtin_mul_overflow(gpt->header.FirstUsableLBA, old_sector_size, &first_usable_byte));
+    first_usable_byte = round_to_next_sector(first_usable_byte, sector_size);
+    if (first_usable_byte < end_of_primary_gpt_table)
+        first_usable_byte = end_of_primary_gpt_table;
+
+    /* Do not bother rounding this.  The division (by sector_size) will do that
+     * anyway. */
+    uint64_t first_unusable_byte = (gpt->header.LastUsableLBA + 1) * (uint64_t)old_sector_size;
+    if (first_unusable_byte > start_of_backup_gpt_table)
+        first_unusable_byte = start_of_backup_gpt_table;
+    memset(new_gpt, 0, sector_size);
+    bool primary = sector_offset == 1;
+    new_gpt->header = (struct GPTHeader){
+        .Signature = {'E', 'F', 'I', ' ', 'P', 'A', 'R', 'T'},
+        .Revision = gpt->header.Revision,
+        .HeaderSize = gpt->header.HeaderSize,
+        .HeaderCRC32 = 0,
+        .Reserved = 0,
+        .MyLBA = sector_offset,
+        .AlternateLBA = primary ? sectors - 1 : 1,
+        .FirstUsableLBA = first_usable_byte / sector_size,
+        .LastUsableLBA = first_unusable_byte / sector_size - 1,
+        .DiskGUID = gpt->header.DiskGUID,
+        .PartitionEntryLBA = primary ? 2 : sectors - 1 - sectors_required,
+        .NumberOfPartitionEntries = gpt->used_entries,
+        .SizeOfPartitionEntry = gpt->header.SizeOfPartitionEntry,
+        .PartitionEntryArrayCRC32 = 0,
+    };
+
+    fprintf(stdout, "LBA is 0x%" PRIX64 ", Alternate LBA is 0x%" PRIX64 ", allocating 0x%" PRIX64 " bytes\n",
+            sector_offset, new_gpt->header.AlternateLBA, bytes_required);
+    fflush(stdout);
+    new_gpt->used_entries = gpt->used_entries;
+
+    new_gpt->entries = aligned_alloc(sector_size, bytes_required);
+    if (new_gpt->entries == NULL) {
+        err(1, "aligned_alloc(%" PRIu32 ", %" PRIu64 ")", sector_size, bytes_required);
+    }
+    for_each_gpt_entry (i, size, count, gpt, entry) {
+        struct GPTPartitionEntry *new_entry = (struct GPTPartitionEntry *)((char *)new_gpt->entries + i * size);
+        memcpy(new_entry, entry, sizeof(*entry));
+        if (!unused_entry(entry)) {
+            uint64_t start_byte_offset = entry->StartingLBA * old_sector_size;
+            uint64_t end_byte_offset = (entry->EndingLBA + 1) * old_sector_size;
+            new_entry->StartingLBA = start_byte_offset / sector_size;
+            new_entry->EndingLBA = end_byte_offset / sector_size - 1;
+        }
+    }
+    *allocated = bytes_required;
+    fixup_gpt(new_gpt, bytes_required);
+    CHECK(new_gpt->entries);
```

Comment:
This, if anything, should be an assert, as you do have proper error checking for this earlier already.